buildSrc/src/main/java/org/elasticsearch/gradle/PublishPlugin.java

```diff
@@ -0,0 +1,202 @@
+                        String artifactId = getArchivesBaseName(projectDependency.getDependencyProject());
+                        dependencyNode.appendNode("artifactId", artifactId);
+                        dependencyNode.appendNode("version", dependency.getVersion());
+                        dependencyNode.appendNode("scope", "compile");
+                    }
+                });
+            });
+        });
+
+        // Add git origin info to generated POM files
+        publishing.getPublications().withType(MavenPublication.class, publication -> {
+            publication.getPom().withXml(PublishPlugin::addScmInfo);
+
+            // have to defer this until archivesBaseName is set
+            project.afterEvaluate(p -> publication.setArtifactId(getArchivesBaseName(project)));
+
+            generatePomTask.configure(
+                t -> t.dependsOn(String.format("generatePomFileFor%sPublication", Util.capitalize(publication.getName())))
+            );
+        });
+
+    }
+
+    private static void addScmInfo(XmlProvider xml) {
+        Node root = xml.asNode();
+        root.appendNode("url", Util.urlFromOrigin(BuildParams.getGitOrigin()));
+        Node scmNode = root.appendNode("scm");
+        scmNode.appendNode("url", BuildParams.getGitOrigin());
+    }
+
+    private static void configureJavadoc(Project project) {
+        // remove compiled classes from the Javadoc classpath: http://mail.openjdk.java.net/pipermail/javadoc-dev/2018-January/000400.html
+        final List<File> classes = new ArrayList<>();
+        project.getTasks().withType(JavaCompile.class).configureEach(javaCompile -> classes.add(javaCompile.getDestinationDir()));
+        project.getTasks().withType(Javadoc.class).configureEach(javadoc -> {
+            // only explicitly set javadoc executable if compiler JDK is different from Gradle
+            // this ensures better cacheability as setting ths input to an absolute path breaks portability
+            Path compilerJvm = BuildParams.getCompilerJavaHome().toPath();
+            Path gradleJvm = Jvm.current().getJavaHome().toPath();
+            try {
+                if (Files.isSameFile(compilerJvm, gradleJvm) == false) {
+                    javadoc.setExecutable(compilerJvm.resolve("bin/javadoc").toString());
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+            javadoc.setClasspath(javadoc.getClasspath().filter(f -> classes.contains(f) == false));
+            /*
+             * Generate docs using html5 to suppress a warning from `javadoc`
+             * that the default will change to html5 in the future.
+             */
+            CoreJavadocOptions javadocOptions = (CoreJavadocOptions) javadoc.getOptions();
+            javadocOptions.addBooleanOption("html5", true);
+        });
+        // ensure javadoc task is run with 'check'
+        project.getTasks()
```

Comment:
Perhaps this bit we leave in the publish plugin, as we don't necessarily want to fail builds for javadoc errors on unpublished modules.

Reply:
I think we do still want this in those cases. If we are writing javadocs, we should make sure they are valid.